### PR TITLE
Simplify render tests

### PR DESCRIPTION
**Pull request by HarelM**
## Launch Checklist

This is meant to convert render tests to jest.
After playing with this I've seen that the performance of jest is significantly slower.
Also for some reason half of the test are failing which probably indicate of a bad refactoring on my part.
I've publishing this WIP PR in order to allow working on this together.
So moving forward I'll improve the harness code and make sure it will continue working and if it won't break again and won't be significantly slower I'll also migrate it to jest.  

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!

### CODE REVIEW

**Comment by HarelM** [comment]
After playing with this a lot I came to the conclusion that jest tests are just too slow to run this and it takes too much time to run this locally.
I removed all the ignored tests as I don't think anyone will fix them.
I moved all the code of the suite implementation, harness, and comparison to a single file.
I removed the server and instead used plain filesystem code.
I added a lot of typings to allow easier writing.
I added some comments above the code for readability.

File: test/integration/render/mock_browser_for_node.ts [hunk 1]
```diff
@@ -1,8 +1,12 @@
+import path, {dirname} from 'path';
+import fs from 'fs';
 import gl from 'gl';
 import {JSDOM, VirtualConsole} from 'jsdom';
 import {PNG} from 'pngjs';
-import request from 'request';
-import {fakeServer} from 'nise';
+import {fileURLToPath} from 'url';
+import '../../unit/lib/web_worker_mock';
+// @ts-ignore
```
Comment (wipfli) [inline test/integration/render/mock_browser_for_node.ts:8]: Is this ts ignore still necessary?

**Review by wipfli — COMMENTED** [review]
Hi Harel. I just skimmed a bit though the code. Thanks for cleaning this up. What is the best way to review this pull request?

File: test/integration/render/mock_browser_for_node.ts (hunk 1 above)
Comment (HarelM) [reply test/integration/render/mock_browser_for_node.ts:8]: According to my IDE, yes.

**Comment by HarelM** [comment]
I would say that the following is probably the most important file in this PR:
`test/integration/render/render.test.ts`.
I would advise to review it and see that you can understand what it does and how, I've tried to add comments in key places.
The rest of the changed files are small changes and the deleted files are not interesting as they are no longer needed.

**Comment by HarelM** [comment]
I've change the threshold to 0.00025 to allow the tests to pass... I don't know why I can't update the images... In any case, I think this is ready for review.

**Comment by wipfli** [comment]
Is render.test.ts a jest test?

**Comment by HarelM** [comment]
No, plain typescript. Jest adds too much overhead and is too slow.

File: test/integration/render/render.test.ts [hunk 2]
```diff
@@ -1,11 +1,574 @@
...
+    }
+    printProgress(style.metadata.test, testStyles.length, ++index);
+}
+
+if (process.env.UPDATE) {
+    console.log(`Updated ${testStyles.length} tests.`);
+    process.exit(0);
+}
+
+const succuess = printStatistics(testStyles.map(s => s.metadata.test));
```
[the remark this answers is not in the document]
Comment (HarelM) [reply test/integration/render/render.test.ts]: Fixed, thanks.

File: test/integration/lib/resultsTemplate.ts [hunk 3]
```diff
@@ -1,70 +0,0 @@
-// eslint-disable-next-line no-unused-expressions
```
Comment (wipfli) [inline test/integration/lib/resultsTemplate.ts:1]: Why is this file not needed anymore?

Comment (HarelM) [reply test/integration/lib/resultsTemplate.ts:1]: I've removed the code to generate html file as I see little use to it - if the tests are failing you need to investigate locally in any case.

File: test/integration/render/suite_implementation.ts [hunk 4]
```diff
@@ -1,246 +0,0 @@
...
-        if (options.collisionDebug) {
-            map.showCollisionBoxes = true;
-            if (options.operations) {
-                options.operations.push(['wait']);
-            } else {
-                options.operations = [['wait']];
-            }
-        }
-        applyOperations(map, options.operations, () => {
-            const viewport = gl.getParameter(gl.VIEWPORT);
```
Comment (wipfli) [inline test/integration/render/suite_implementation.ts:124]: Were there two `applyOperations` functions?

File: test/integration/render/tests/zoomed-fill/negative-zoom/style.json [hunk 5]
```diff
@@ -2,7 +2,6 @@
   "version": 8,
   "metadata": {
     "test": {
-      "allowed": 0.0002,
```
Comment (wipfli) [inline test/integration/render/tests/zoomed-fill/negative-zoom/style.json:5]: Please comment?

File: test/integration/render/suite_implementation.ts (hunk 4 above)
Comment (HarelM) [reply test/integration/render/suite_implementation.ts:124]: For render test, just this one.

File: test/integration/render/tests/zoomed-fill/negative-zoom/style.json (hunk 5 above)
Comment (HarelM) [reply test/integration/render/tests/zoomed-fill/negative-zoom/style.json:5]: I removed the special case as now all tests use 0.00025 and this continued to fail due to github env change (fixes #1008 as well)

File: test/integration/render/render.test.ts [hunk 6]
```diff
@@ -1,11 +1,574 @@
...
+        return defaultValue;
+
+    const split = String(options.tests.splice(index, 1)).split('=');
+    if (split.length !== 2)
+        return defaultValue;
+
+    return split[1];
+}
+/**
+ * This method compares the buffer that was created to the expected file in the file system.
```
Comment (wipfli) [inline test/integration/render/render.test.ts:113]: Remove "This method". Use "Compares the buffer..."

File: test/integration/render/render.test.ts [hunk 7]
```diff
@@ -1,11 +1,574 @@
...
+
+    fs.writeFileSync(diffPath, diffBuf);
+    fs.writeFileSync(actualPath, actualBuf);
+
+    testData.difference = minDiff;
+    testData.ok = minDiff <= testData.allowed;
+}
+
+/**
+ * This function mocks XHR request and simply pulls file from the file system.
```
Comment (wipfli) [inline test/integration/render/render.test.ts:194]: Remove "This function"

File: test/integration/render/render.test.ts [hunk 8]
```diff
@@ -1,11 +1,574 @@
...
+            } catch (ex) {
+                req.setStatus(404); // file not found
+                req.onload();
+            }
+        }, 0);
+    };
+}
+
+/**
+ * This method gets all the tests from the file system looking for style.json files.
```
Comment (wipfli) [inline test/integration/render/render.test.ts:229]: Remove "This method"

File: test/integration/render/render.test.ts [hunk 9]
```diff
@@ -1,11 +1,574 @@
...
+
+    if (erroredCount > 0) {
+        console.log('%d errored (%s%)',
+            erroredCount, (100 * erroredCount / totalCount).toFixed(1));
+    }
+
+    return (failedCount + erroredCount) === 0;
+}
+
+/**
```
Comment (wipfli) [inline test/integration/render/render.test.ts:533]: I don't know if `/**` shoulf be exclusively ude with JSDocs...

**Review by wipfli — APPROVED** [review]
I checked the code on a high level and it looks like all that was there before is still around. You can do the small improvements if you would like to. I assume all deleted tests were ignored ones.

**Comment by HarelM** [comment]
Thanks for the review!! I'll do the fixes and merge them later tonight.

File: test/integration/render/render.test.ts (hunk 9 above)
Comment (HarelM) [reply test/integration/render/render.test.ts:533]: I think this is OK in this case as this is a long comment and using `//` in this case is a bit of a hassle...

**Comment by birkskyum** [comment]
Awesome job with this! I do get this warning when I try to run the tests:

TypeError [ERR_IMPORT_ASSERTION_TYPE_MISSING]: Module "./src/style-spec/reference/v8.json" needs an import assertion of type "json"

Have you seen that before while working on this?

**Comment by birkskyum** [comment]
Downgraded to 16.13, and then everything works.

**Comment by HarelM** [comment]
Yes, there're breaking changes in 16.14 that cause this. I haven't fixed them yet, seems that typescript is not very fond of the fix...
All render pass in CI, some do not pass locally on my mac though.

**Pull request merged**